core/primitives/src/utils/io.rs

```diff
@@ -0,0 +1,182 @@
+use std::io::{self, Read, Write};
+
+/// Wrapper for Write that counts number of bytes written.
+/// It also accepts an optional limit for the total number of bytes written;
+/// if this limit is exceeded, write operation raises an io::Error of kind WriteZero.
+pub struct CountingWrite<W: Write> {
+    inner: W,
+    /// Total number of bytes written.
+    total_written: u64,
+    /// If present, the total number of bytes allowed to be written.
+    /// If this limit is reached, the wirte operation will return an error.
+    limit: Option<u64>,
```

Comment:
A micro-optimization, but it wouldn't be a terrible idea to use a sentinel value for limit to mean "no limit". In particular `u64::MAX` would be a suitable value, since it can't conceivably be reached. This would remove a branch from the implementation of `read` and `write` calls, reduce the size of this structure and somewhat push the wrapper itself towards having a single responsibility (as it would no longer "*…and optionally limits*" case.)

Reply:
Like it. Done.